### PR TITLE
Binary!

### PR DESCRIPTION
**Pull request by ChrisMacNaughton**
This adds in a basic binary that will convert between json and binary formats for the `CrushMap`

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -87,6 +90,15 @@ pub enum EncodingError {
...
 }
 
+#[derive(Debug)]
+pub enum CephVersion {
+    Argonaut,
+    Bobtail,
+    Firefly,
+    Hammer,
+    Jewel,
+}
```
Comment (cholcombe973) [inline src/lib.rs:100]: Are we going to support the intermediate versions?

File: src/lib.rs [hunk 2]
```diff
@@ -207,6 +219,30 @@ pub enum BucketTypes {
...
+            BucketTypes::List(ref b) => Some(&b.bucket),
+            BucketTypes::Tree(ref b) => Some(&b.bucket),
+            BucketTypes::Straw(ref b) => Some(&b.bucket),
+            BucketTypes::Straw2(ref b) => Some(&b.bucket),
+        }
+    }
+
+    pub fn id(&self) -> i32 {
+        match *self {
+            BucketTypes::Unknown => 65536,
```
Comment (cholcombe973) [inline src/lib.rs:236]: this could be set to i32::max_value()

File: src/lib.rs [hunk 3]
```diff
@@ -307,3 +343,97 @@ pub struct CrushMap {
     /// device fails.
     pub chooseleaf_stable: Option<u8>,
 }
+
+impl CrushMap {
+    pub fn with_tunables(mut self, version: CephVersion) -> Self {
```
Comment (cholcombe973) [inline src/lib.rs:348]: I like this!  Nice

File: src/lib.rs [hunk 4]
```diff
@@ -307,3 +343,97 @@ pub struct CrushMap {
...
+            chooseleaf_vary_r: Some(0),
+            straw_calc_version: Some(0),
+            allowed_bucket_algorithms: Some(0),
+            chooseleaf_stable: Some(22),
+        }
+    }
+}
+#[derive(Clone, Debug, Eq, PartialEq, RustcDecodable, RustcEncodable)]
+pub struct CephDisk {
+    pub name: Option<String>,
```
Comment (cholcombe973) [inline src/lib.rs:411]: Why are these fields optional?  I would think they shouldn't be

File: src/lib.rs [hunk 5]
```diff
@@ -307,3 +343,97 @@ pub struct CrushMap {
...
+    pub uuid: Option<Uuid>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, RustcDecodable, RustcEncodable)]
+pub struct CephHost {
+    pub hostname: Option<String>,
```
Comment (cholcombe973) [inline src/lib.rs:417]: Same here.  If you make a host shouldn't be required that it have a name?  I can see disks being optional if the host will be used for rgw or something

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+//                 CephBucketType};
+
+arg_enum!{
+  enum Mode {
+    compile,
+    decompile
+  }
+}
+
+fn main() {
```
Comment (cholcombe973) [inline src/main.rs:29]: Just a general comment, could you remove the unwraps so we fail cleaner?

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            .takes_value(true))
+        .get_matches();
+
+
+    let mode = value_t!(matches.value_of("mode"), Mode).unwrap();
+
+    match mode {
+        Mode::compile => {
+            let mut input = String::new();
+            io::stdin().read_to_string(&mut input).unwrap();
```
Comment (cholcombe973) [inline src/main.rs]: This is probably fine.  I doubt we'll ever get a crushmap input that is bigger than the heap size on the machine.

**Review by cholcombe973 — CHANGES_REQUESTED** [review]
I like where this is going.

File: src/lib.rs (hunk 1 above)
Comment (ChrisMacNaughton) [reply src/lib.rs:100]: These are currently only used for setting the tunables

File: src/lib.rs (hunk 4 above)
Comment (ChrisMacNaughton) [reply src/lib.rs:411]: We can change that but the crushmap itself has them as optional

**Pull request merged**